src/hotspot/share/utilities/vmError.cpp

```diff
@@ -166,6 +170,48 @@ static void print_bug_submit_message(outputStream *out, Thread *thread) {
   out->print_raw_cr("#");
 }
 
+static bool check_stack_headroom(Thread* thread,
+                                 size_t headroom) {
+  static const address stack_top    = thread != nullptr
+                                    ? thread->stack_base()
+                                    : os::current_stack_base();
+  static const size_t  stack_size   = thread != nullptr
+                                    ? thread->stack_size()
+                                    : os::current_stack_size();
+  static const address stack_bottom = stack_top - stack_size;
+
+  const address stack_pointer = os::current_stack_pointer();
+
+  const ptrdiff_t stack_headroom = stack_pointer - stack_bottom;
+  return (stack_pointer < stack_bottom || stack_headroom < 0 ||
+          static_cast<size_t>(stack_headroom) < headroom);
+}
+
+#ifdef ASSERT
+void VMError::reenterant_test_hit_stack_limit() {
+  if (!check_stack_headroom(_thread, _reattempt_required_stack_headroom)) {
+    char stack_buffer[_reattempt_required_stack_headroom / 2];
+    static_cast<void>(stack_buffer[sizeof(stack_buffer) - 1] = '\0');
+    reenterant_test_hit_stack_limit();
+  }
+  controlled_crash(14);
+}
+#endif // ASSERT
+
+bool VMError::should_stop_reattempt_step(const char* &reason) {
```

Comment:
I had to read this twice to see the "stop" in the name :-) 

I would prefer the logic to be inverse and this function to be named "can_reattempt_step". But since this is a matter of taste, I leave it up to you.